File: src/test/java/PuzzleMasterTest.java
import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

public class PuzzleMasterTest {

    private static PuzzleMaster createSmallInstance() {
        final List<String> words = Arrays.asList(
                "abracadabrazy",
                "abrac",
                "barca",
                "barbar",
                "zzzzz",
                "zzzzzzzz",
                "lengthened",
                "lengthen",
                "then",
                "vis-a-vis",
                "Caps"
        );
        return new PuzzleMaster(words);
    }

    @Test
    public void constructor_onSmallData_createsWords() {
        Assert.assertEquals(
                new HashSet<>(Arrays.asList(
                        "abracadabrazy",
                        "abrac",
                        "barca",
                        "barbar",
                        "zzzzz",
                        "zzzzzzzz",
                        "lengthened",
                        "lengthen")),
                createSmallInstance().words);
    }

    @Test
    public void constructor_onSmallData_createsPots() {
        Assert.assertEquals(
                new HashSet<>(Arrays.asList(
                        Puzzle.characterVector("abrcdzy".toCharArray()),
                        Puzzle.characterVector("lengthd".toCharArray()))),
                createSmallInstance().pots);
    }

    @Test
    public void constructor_onSmallData_createsWordsByVector() {
        final PuzzleMaster pm = createSmallInstance();
        Assert.assertEquals(new HashSet<>(Arrays.asList(
                Puzzle.characterVector("abrcdzy".toCharArray()),
                Puzzle.characterVector("abrc".toCharArray()),
                Puzzle.characterVector("bar".toCharArray()),
                Puzzle.characterVector("z".toCharArray()),
                Puzzle.characterVector("lengthd".toCharArray()),
                Puzzle.characterVector("length".toCharArray()))
        ), pm.wordsByVector.keySet());
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("abracadabrazy")),
                pm.wordsByVector.get(Puzzle.characterVector("abrcdzy".toCharArray())));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("abrac", "barca")),
                pm.wordsByVector.get(Puzzle.characterVector("abrc".toCharArray())));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("barbar")),
                pm.wordsByVector.get(Puzzle.characterVector("bar".toCharArray())));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("zzzzz", "zzzzzzzz")),
                pm.wordsByVector.get(Puzzle.characterVector("z".toCharArray())));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("lengthened")),
                pm.wordsByVector.get(Puzzle.characterVector("lengthd".toCharArray())));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("lengthen")),
                pm.wordsByVector.get(Puzzle.characterVector("length".toCharArray())));
    }

    @Test
    public void constructor_onSmallData_createsPuzzles() {
        final PuzzleMaster pm = createSmallInstance();
        Assert.assertEquals(14, pm.puzzles.size());
        for (String pot : new String[] { "abrcdzy", "lengthd" }) {
            for (char required : pot.toCharArray()) {
                final Puzzle p = new Puzzle(
                        Puzzle.characterVector(pot.toCharArray()),
                        Puzzle.characterVector(new char[] { required }));
                Assert.assertTrue(p.toString(), pm.puzzles.contains(p));
            }
        }
    }

    @Test
    public void solutionsTo_simple1() {
        final PuzzleMaster pm = createSmallInstance();
        final Puzzle puzzle = new Puzzle(
                Puzzle.characterVector("abrcdzy".toCharArray()),
                Puzzle.characterVector("c".toCharArray()));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("abracadabrazy", "abrac", "barca")),
                pm.solutionsTo(puzzle));
    }

    public void solutionsTo_simple2() {
        final PuzzleMaster pm = createSmallInstance();
        final Puzzle puzzle = new Puzzle(
                Puzzle.characterVector("lengthd".toCharArray()),
                Puzzle.characterVector("e".toCharArray()));
        Assert.assertEquals(
                new HashSet<>(Arrays.asList("lengthened", "lengthen")),
                pm.solutionsTo(puzzle));
    }

    @Test
    public void solutionsTo_noSolutions() {
        final PuzzleMaster pm = createSmallInstance();
        final Puzzle puzzle = new Puzzle(
                Puzzle.characterVector("abcdleg".toCharArray()),
                Puzzle.characterVector("c".toCharArray()));
        Assert.assertEquals(Collections.emptySet(), pm.solutionsTo(puzzle));
    }

    @Test
    public void solutionsTo_unknownLetters() {
        final PuzzleMaster pm = createSmallInstance();
        final Puzzle puzzle = new Puzzle(
                Puzzle.characterVector("jkopqsu".toCharArray()),
                Puzzle.characterVector("q".toCharArray()));
        for (String word : pm.words) {
            Assert.assertTrue("test data mismatch",
                    (puzzle.potVector & Puzzle.characterVector(word.toCharArray())) == 0);
        }
        Assert.assertEquals(Collections.emptySet(), pm.solutionsTo(puzzle));
    }

}

File: src/puzzleMaster.test.js
// @flow

import type {CharVec} from "./charvec";
import {stringToVector, vectorToString} from "./charvec";
import PuzzleMaster from "./puzzleMaster";

describe("PuzzleMaster", () => {
  const words = () => [
    "abracadabrazy",
    "abrac",
    "barca",
    "barbar",
    "zzzzz",
    "zzzzzzzz",
    "lengthened",
    "lengthen",
    "then", // too short
    "vis-a-vis", // outside the alphabet
    "Caps", // outside the alphabet
    "abcdefgh", // too many distinct letters
  ];

  describe("constructor", () => {
    it("filters out invalid words", () => {
      const instance = new PuzzleMaster(words());
      expect(instance.words.slice().sort()).toEqual(
        [
          "abracadabrazy",
          "abrac",
          "barca",
          "barbar",
          "zzzzz",
          "zzzzzzzz",
          "lengthened",
          "lengthen",
        ].sort()
      );
    });

    it("identifies pots", () => {
      const instance = new PuzzleMaster(words());
      expect(instance.pots).toEqual(
        new Set([stringToVector("abrcdzy"), stringToVector("lengthd")])
      );
    });

    it("partitions words by vector", () => {
      function sorted<T>(xs: Iterable<T>): T[] {
        return Array.from(xs)
          .slice()
          .sort();
      }
      function really<T>(value: ?T): T {
        if (value == null) {
          throw new Error(String(value));
        } else {
          return value;
        }
      }
      const instance = new PuzzleMaster(words());
      const actual: Map<CharVec, string[]> = instance.wordsByVector;
      const expected: Map<CharVec, string[]> = new Map([
        [stringToVector("abrcdzy"), ["abracadabrazy"]],
        [stringToVector("abrc"), ["abrac", "barca"]],
        [stringToVector("bar"), ["barbar"]],
        [stringToVector("z"), ["zzzzz", "zzzzzzzz"]],
        [stringToVector("lengthd"), ["lengthened"]],
        [stringToVector("length"), ["lengthen"]],
      ]);
      expect(sorted(actual.keys())).toEqual(sorted(expected.keys()));
      for (const key of expected.keys()) {
        expect(sorted(really(actual.get(key)))).toEqual(
          sorted(really(expected.get(key)))
        );
      }
    });
  });

  describe("#solve", () => {
    it("solves a simple puzzle", () => {
      const instance = new PuzzleMaster(words());
      const puzzle = {
        required: stringToVector("c"),
        optional: stringToVector("abrdzy"),
      };
      const actual = instance.solutionsTo(puzzle);
      const expected = new Set(["abracadabrazy", "abrac", "barca"]);
      expect(actual).toEqual(expected);
    });

    it("solves another simple puzzle", () => {
      const instance = new PuzzleMaster(words());
      const puzzle = {
        required: stringToVector("e"),
        optional: stringToVector("lngthd"),
      };
      const actual = instance.solutionsTo(puzzle);
      const expected = new Set(["lengthened", "lengthen"]);
      expect(actual).toEqual(expected);
    });

    it("solves a puzzle with no solutions", () => {
      const instance = new PuzzleMaster(words());
      const puzzle = {
        required: stringToVector("c"),
        optional: stringToVector("abdleg"),
      };
      const actual = instance.solutionsTo(puzzle);
      const expected = new Set([]);
      expect(actual).toEqual(expected);
    });

    it("solves a puzzle with never-before-seen letters", () => {
      const instance = new PuzzleMaster(words());
      const puzzle = {
        required: stringToVector("q"),
        optional: stringToVector("jkouwx"),
      };

      // Self-check...
      {
        const pot = puzzle.required | puzzle.optional;
        for (const word of words()) {
          const overlap = stringToVector(word) & pot;
          if (overlap) {
            const overlapString = vectorToString(overlap);
            throw new Error(
              `word "${word}" contains forbidden letters: "${overlapString}"`
            );
          }
        }
      }

      const actual = instance.solutionsTo(puzzle);
      const expected = new Set([]);
      expect(actual).toEqual(expected);
    });
  });
});
